Sources/MapboxCoreNavigation/RouteProgress.swift

```diff
@@ -10,61 +10,82 @@ open class RouteProgress: Codable {
     private static let reroutingAccuracy: CLLocationAccuracy = 90
     
     /**
-     Returns the current `Route`.
+     Intializes a new `RouteProgress`.
+
+     - parameter route: The route to follow.
+     - parameter options: The route options that were attached to the route request.
+     - parameter legIndex: Zero-based index indicating the current leg the user is on.
      */
-    public var route: Route
+    public init(route: Route, options: RouteOptions, legIndex: Int = 0, spokenInstructionIndex: Int = 0) {
+        self.route = route
+        self.routeOptions = options
+        self.legIndex = legIndex
+        self.currentLegProgress = RouteLegProgress(leg: route.legs[legIndex], stepIndex: 0, spokenInstructionIndex: spokenInstructionIndex)
+
+        self.calculateLegsCongestion()
+    }
     
-    public let routeOptions: RouteOptions
+    func reroutingOptions(with current: CLLocation) -> RouteOptions {
+        let oldOptions = routeOptions
+        let user = Waypoint(coordinate: current.coordinate)
 
-    /**
-     Index representing current `RouteLeg`.
-     */
-    public var legIndex: Int {
-        didSet {
-            assert(legIndex >= 0 && legIndex < route.legs.endIndex)
-            // TODO: Set stepIndex to 0 or last index based on whether leg index was incremented or decremented.
-            currentLegProgress = RouteLegProgress(leg: currentLeg)
-            
-            legIndexHandler?(oldValue, legIndex)
+        if (current.course >= 0) {
+            user.heading = current.course
+            user.headingAccuracy = RouteProgress.reroutingAccuracy
         }
-    }
-    typealias LegIndexHandlerAction = (_ oldValue: Int, _ newValue: Int) -> ()
-    var legIndexHandler: LegIndexHandlerAction?
-    /**
-     If waypoints are provided in the `Route`, this will contain which leg the user is on.
-     */
-    public var currentLeg: RouteLeg {
-        return route.legs[legIndex]
-    }
+        let newWaypoints = [user] + remainingWaypointsForCalculatingRoute()
+        let newOptions = oldOptions.copy() as! RouteOptions
+        newOptions.waypoints = newWaypoints
 
-    /**
-     Returns the remaining legs left on the current route
-     */
-    public var remainingLegs: [RouteLeg] {
-        return Array(route.legs.suffix(from: legIndex + 1))
+        return newOptions
     }
-
+    
+    // MARK: Interacting with Route
```

Comment:
```suggestion
    // MARK: Details About the Route
```

I think we should split off the methods into a separate mark called “Updating the Route Progress” and the computed properties into a separate mark called “Route Statistics”.